### ISSUE TITLE
Training two models in parallel multithreaded does not scale unless you make it two different programs

### ISSUE THREAD

**Issue by IvoTops** [Microsoft/CNTK#2662]
Using C# API, CNTK 2.3 GPU on Windows 10
_______________________________________________________________________________________
Training one model -> 60 epochs/minute  (GPULoad 30%)
Training two models multithreaded, no locking -> 60 epochs/minute combined, 30 for each  (GPULoad 30%)

Running the Training as two different programs at the same time
gives 110 epochs/minute combined (GPULoad 55%)

Running the Training as three different programs at the same time
gives 140 epochs/minute combined (GPULoad 70%)

For the multithreaded version I tried with and without a device lock, makes no difference.
_________________________________________________________________________________________
Basically to scale out I will have to spawn a new process for every model. I think the API should be able to scale without this.

**Comment by rhy-ama** [comment]
Is 2.3 version now not thread safe and threading is implicitly handled by C# API? 

What design are you using to launch training in parallel. You can encapsulate you CNTK API calls within an object instance and launching training processes via task[id]=Task.Factory.StartNew( ()=>yourObject[id].run()) / WaitAll(tasks). This is what we use and it scales with little to none overhead (yet to be tested on 2.3)

**Comment by IvoTops** [comment]
@rhy-ama If you say your app scales without virtually overhead are you talking about doing multiple parallel TrainMiniBatch calls (on different models/trainers) all on one single GPU?

Because my profiler tells my all my threads are 100% in the TrainMiniBatch API call. So I really think that call has some serious blocking inside. It only really executes one thread at the time. As opposed to running the threads as seperate programs.
